File: OCDemo/OCDemo/设计模式/DesignModeViewController.m
//
//  DesignModeViewController.m
//  OCDemo
//
//  Created by mkrq-yh on 2019/3/19.
//  Copyright © 2019 mkrq-yh. All rights reserved.
//

#import "DesignModeViewController.h"
#import "KVOViewController.h"
#import "NotificationViewController.h"
#import "DelegateViewController.h"
#import "FactoryViewController.h"

@interface DesignModeViewController ()


@end

@implementation DesignModeViewController


- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.navigationItem.title = @"设计模式";
    self.view.backgroundColor = [UIColor whiteColor];
    
    UIScrollView *scrollView = [[UIScrollView alloc] initWithFrame:self.view.bounds];
    [self.view addSubview:scrollView];
    
    UIButton *btn1 = [UIButton buttonWithType:UIButtonTypeCustom];
    btn1.frame = CGRectMake(30, 50, self.view.frame.size.width-60, 50);
    [btn1 setTitle:@"KVO设计模式" forState:UIControlStateNormal];
    [btn1 setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    btn1.titleLabel.font = [UIFont systemFontOfSize:16];
    btn1.backgroundColor = [UIColor greenColor];
    [btn1 addTarget:self action:@selector(kvoModeBtnLClick) forControlEvents:UIControlEventTouchUpInside];
    [scrollView addSubview:btn1];
    
    UIButton *btn2 = [UIButton buttonWithType:UIButtonTypeCustom];
    btn2.frame = CGRectMake(30, btn1.frame.origin.y+btn1.frame.size.height+20, self.view.frame.size.width-60, 50);
    [btn2 setTitle:@"通知设计模式" forState:UIControlStateNormal];
    [btn2 setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    btn2.titleLabel.font = [UIFont systemFontOfSize:16];
    btn2.backgroundColor = [UIColor greenColor];
    [btn2 addTarget:self action:@selector(notifiModeBtnLClick) forControlEvents:UIControlEventTouchUpInside];
    [scrollView addSubview:btn2];
    
    UIButton *btn3 = [UIButton buttonWithType:UIButtonTypeCustom];
    btn3.frame = CGRectMake(30, btn2.frame.origin.y+btn2.frame.size.height+20, self.view.frame.size.width-60, 50);
    [btn3 setTitle:@"代理设计模式" forState:UIControlStateNormal];
    [btn3 setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    btn3.titleLabel.font = [UIFont systemFontOfSize:16];
    btn3.backgroundColor = [UIColor greenColor];
    [btn3 addTarget:self action:@selector(delegateModeBtnLClick) forControlEvents:UIControlEventTouchUpInside];
    [scrollView addSubview:btn3];
    
    UIButton *btn4 = [UIButton buttonWithType:UIButtonTypeCustom];
    btn4.frame = CGRectMake(30, btn3.frame.origin.y+btn3.frame.size.height+20, self.view.frame.size.width-60, 50);
    [btn4 setTitle:@"工厂设计模式" forState:UIControlStateNormal];
    [btn4 setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    btn4.titleLabel.font = [UIFont systemFontOfSize:16];
    btn4.backgroundColor = [UIColor greenColor];
    [btn4 addTarget:self action:@selector(factoryModeBtnLClick) forControlEvents:UIControlEventTouchUpInside];
    [scrollView addSubview:btn4];
    
}

- (void)kvoModeBtnLClick {
    KVOViewController *mode = [[KVOViewController alloc] init];
    [self.navigationController pushViewController:mode animated:true];
}

- (void)notifiModeBtnLClick {
    NotificationViewController *mode = [[NotificationViewController alloc] init];
    [self.navigationController pushViewController:mode animated:true];
}

- (void)delegateModeBtnLClick {
    DelegateViewController *mode = [[DelegateViewController alloc] init];
    [self.navigationController pushViewController:mode animated:true];
}

- (void)factoryModeBtnLClick {
    FactoryViewController *mode = [[FactoryViewController alloc] init];
    [self.navigationController pushViewController:mode animated:true];
}


/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SwiftDemo/SwiftDemo/设计模式/DesignModeViewController.swift

//
//  DesignModeViewController.swift
//  SwiftDemo
//
//  Created by mkrq-yh on 2019/3/20.
//  Copyright © 2019 mkrq-yh. All rights reserved.
//

import UIKit

class DesignModeViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.white
        self.navigationItem.title = "设计模式"
        
        let kvoBtn = UIButton(type: .custom)
        kvoBtn.frame = CGRect(x: 20, y: 100, width: kScreenWidth-40, height: 40)
        kvoBtn.backgroundColor = UIColor.green
        kvoBtn.setTitle("KVO设计模式", for: .normal)
        kvoBtn.setTitleColor(UIColor.red, for: .normal)
        kvoBtn.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        kvoBtn.addTarget(self, action: #selector(kvoBtnClick(sender:)), for: .touchUpInside)
        self.view.addSubview(kvoBtn)
        
        let nitifiBtn = UIButton(type: .custom)
        nitifiBtn.frame = CGRect(x: 20, y: kvoBtn.frame.origin.y+kvoBtn.frame.size.height+20, width: kScreenWidth-40, height: 40)
        nitifiBtn.backgroundColor = UIColor.green
        nitifiBtn.setTitle("通知", for: .normal)
        nitifiBtn.setTitleColor(UIColor.red, for: .normal)
        nitifiBtn.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        nitifiBtn.addTarget(self, action: #selector(nitifiBtnClick(sender:)), for: .touchUpInside)
        self.view.addSubview(nitifiBtn)
        
        let delegateBtn = UIButton(type: .custom)
        delegateBtn.frame = CGRect(x: 20, y: nitifiBtn.frame.origin.y+nitifiBtn.frame.size.height+20, width: kScreenWidth-40, height: 40)
        delegateBtn.backgroundColor = UIColor.green
        delegateBtn.setTitle("代理", for: .normal)
        delegateBtn.setTitleColor(UIColor.red, for: .normal)
        delegateBtn.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        delegateBtn.addTarget(self, action: #selector(delegateBtnClick(sender:)), for: .touchUpInside)
        self.view.addSubview(delegateBtn)
        
        let factoryBtn = UIButton(type: .custom)
        factoryBtn.frame = CGRect(x: 20, y: delegateBtn.frame.origin.y+delegateBtn.frame.size.height+20, width: kScreenWidth-40, height: 40)
        factoryBtn.backgroundColor = UIColor.green
        factoryBtn.setTitle("工厂", for: .normal)
        factoryBtn.setTitleColor(UIColor.red, for: .normal)
        factoryBtn.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        factoryBtn.addTarget(self, action: #selector(factoryBtnClick(sender:)), for: .touchUpInside)
        self.view.addSubview(factoryBtn)
    }
    
    @objc func kvoBtnClick(sender:UIButton) {
        let kvo = KVOViewController()
        self.navigationController?.pushViewController(kvo, animated: true)
        
    }
    
    @objc func nitifiBtnClick(sender:UIButton) {
        let notifi = NotificationViewController()
        self.navigationController?.pushViewController(notifi, animated: true)
    }
    
    @objc func delegateBtnClick(sender:UIButton) {
        let del = DelegateViewController()
        self.navigationController?.pushViewController(del, animated: true)
    }

    @objc func factoryBtnClick(sender:UIButton) {
        let fac = FactoryViewController()
        self.navigationController?.pushViewController(fac, animated: true)
    }
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
